test/unit/uk/gov/hmrc/organisationsdetailsapi/services/CacheServiceSpec.scala

```diff
@@ -0,0 +1,130 @@
+import org.mockito.ArgumentMatchers.{eq => eqTo, _}
+import org.mockito.Mockito.{verify, verifyNoInteractions}
+import org.scalatestplus.mockito.MockitoSugar
+import play.api.libs.json.{Json, OFormat}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.organisationsdetailsapi.cache.{CacheConfiguration, ShortLivedCache}
+import uk.gov.hmrc.organisationsdetailsapi.services.{CacheIdBase, CacheService, PayeCacheId, SaCacheId}
+import utils.TestSupport
+
+import scala.concurrent.Future
+
+class CacheServiceSpec extends TestSupport with MockitoSugar {
+
+  val cacheId = TestCacheId("foo")
+  val cachedValue = TestClass("cached value")
+  val newValue = TestClass("new value")
+
+  trait Setup {
+
+    val mockClient = mock[ShortLivedCache]
+    val mockCacheConfig = mock[CacheConfiguration]
+    val cacheService = new CacheService {
+      override val shortLivedCache: ShortLivedCache = mockClient
+      override val conf: CacheConfiguration = mockCacheConfig
+      override val key: String = "test"
+    }
+
+    implicit val hc: HeaderCarrier = HeaderCarrier()
+
+    given(mockCacheConfig.cacheEnabled).willReturn(true)
+
+  }
+
+  "cacheService.get" should {
+
+    "return the cached value for a given id and key" in new Setup {
+
+      given(mockClient.fetchAndGetEntry[TestClass](eqTo(cacheId.id), eqTo(cacheService.key))(any()))
+        .willReturn(Future.successful(Some(cachedValue)))
+      await(cacheService.get[TestClass](cacheId, Future.successful(newValue))) shouldBe cachedValue
+
+    }
+
+    "cache the result of the fallback function when no cached value exists for a given id and key" in new Setup {
+
+      given(mockClient.fetchAndGetEntry[TestClass](eqTo(cacheId.id), eqTo(cacheService.key))(any()))
+        .willReturn(Future.successful(None))
+
+      await(cacheService.get[TestClass](cacheId, Future.successful(newValue))) shouldBe newValue
+      verify(mockClient).cache[TestClass](eqTo(cacheId.id), eqTo(cacheService.key), eqTo(newValue))(any())
+
+    }
+
+    "ignore the cache when caching is not enabled" in new Setup {
+
+      given(mockCacheConfig.cacheEnabled).willReturn(false)
+      await(cacheService.get[TestClass](cacheId, Future.successful(newValue))) shouldBe newValue
+      verifyNoInteractions(mockClient)
+
+    }
+  }
+
+  "PayeCacheId" should {
+
+    "produce a cache id based on matchId and scopes" in {
+
+      val matchId = UUID.randomUUID()
+      val fromDateString = "2017-03-02"
+      val toDateString = "2017-05-31"
+
+      val interval = new Interval(
+        new LocalDate(fromDateString).toDateTimeAtStartOfDay,
+        new LocalDate(toDateString).toDateTimeAtStartOfDay)
+
+      val fields = "ABDFH"
+
+      PayeCacheId(matchId, interval, fields).id shouldBe
+        s"$matchId-${interval.getStart}-${interval.getEnd}-ABDFH"
+
+    }
+
+  }
+
+  "SaCacheId" should {
+
+    "produce a cache id based on matchId and scopes" in {
+
+      val matchId = UUID.randomUUID()
```

Comment:
Once the above comments are addressed the cache key will be simplified as it will not longer require the date ranges or the fields parameter. This will need to be refactored to align to the new cache key format.